core/dbt/cli/main.py

```diff
@@ -25,6 +28,52 @@ def cli_runner():
     cli()
 
 
+class dbtUsageException(Exception):
+    pass
+
+
+# Programmatic invocation
+class dbtRunner:
```

Comment:
Is this first rough version of thin wrapper, for end users who want to programmatically invoke `dbt-core` via Python API? Thinking:
```python
    # initialize longer-lived dbtRunner
    # by implicitly loading project/profile/etc based on config + cwd
    dbt = dbtRunner()
    
    # submit commands to dbtRunner
    # each command returns
    #   results: List[Result]
    #   success: bool
    results, success = dbt.run(...)
    results, success = dbt.test(...)
```

Reply:
```python
    results, success = dbt.run(...)
    results, success = dbt.test(...)
```
this part is tbd I think? current way would be pass in command through args, the main advantage is some CLI args need to be specified before the actual command. So exposing things this way provide best flexibility at the moment.

```python
args = ['run', '--project-dir', 'something']
res, success = dbt.invoke(args)
```

Reply:
> current way would be pass in command through args

We'll definitely want to support this generic pattern ("given an arbitrary list of CLI arg strings, invoke"). I could also see us wanting to support a more structured approach, where users can build up a dictionary of attributes and pass it into a named command. All of which is to say, we should design this programmatic interface for real, before we commit to it as a public-facing API — and that means sitting down, writing some user stories, talking with community members. All good stuff to do next year :)

> the main advantage is some CLI args need to be specified before the actual command

Lukewarm take: This is pretty confusing for end users! We should really seek to support flags in any order (and resolve any resulting ambiguities by actually renaming flags). I know we implemented the new CLI modeling in accordance with the old CLI model — parity is a good goal to shoot for with any refactor — but it would be _great_ to just support both of these, seamlessly and equivalently:
```bash
$ dbt --no-use-color run
$ dbt run --no-use-color
```

Reply:
cc @iknox-fa

Reply:
Well, that wasn't part of the original plan, but I guess we can see about supporting arbitrary flag ordering--  It's not something we should take on in this PR though.  @jtcohen6, can you write up a ticket that we can put into the estimation process.

Reply:
@iknox-fa Totally - opened here: https://github.com/dbt-labs/dbt-core/issues/6497